Pd documents directory & deken adds to search paths

danomatika opened the pull request:
This is a PR for a proposed Pd "documents directory" and deken improvements.

The Pd documents directory is:

* a common sense, beginner-friendly location for patches and externals
* a default deken install directory
* an initial directory for file dialogs (ignored when pd is started from the command line)
* a GUI-only preferences value
* configurable in the Paths dialog
* completely optional!

**[See the demo video on Youtube](https://youtu.be/YWMnynyIeoc)**

On startup, Pd checks if the documents directory is valid. If not, it prompts the user about either creating it (first run) or choosing a new location if it is not found. Hitting "No" disables this feature.

<img width="612" alt="screen shot 2017-07-31 at 12 28 19 pm" src="https://user-images.githubusercontent.com/480637/28775194-041d1722-75f1-11e7-832b-bfb7cee9d465.png">

Default location is: `HOME/Documents/Pd` and the fallback is `HOME/Pd`

When created, an "externals" sub directory is also created and set as the deken install path:

As with the above: `HOME/Documents/Pd/externals` or `HOME/Pd/externals`

If `HOME` cannot be determined or is not writable, the documents directory feature is automatically disabled.

The behavior of both the documents directory and the deken install directory can be configured in the Path dialog. Changing the Documents directory also changes the deken install dir by default.

<img width="584" alt="screen shot 2017-07-31 at 12 28 26 pm" src="https://user-images.githubusercontent.com/480637/28775259-4cc2959c-75f1-11e7-888d-eef2145a988c.png">

When set, file open & save dialogs start in the documents directory. Also, when deken asks where to download externals, the documents directory externals subdirectory is used be default and users can override and use a different location. This temporary location is not saved.

Deken improvements include:

* display external name in dialog prompts
* optionally add a downloaded external directory to the search paths

<img width="782" alt="screen shot 2017-07-31 at 12 30 13 pm" src="https://user-images.githubusercontent.com/480637/28775286-6ab625aa-75f1-11e7-85f5-6580115473e3.png">

Adding to the search paths allows for quick usage of declare without the second step of adding paths by hand. This may not handle all situations (ie. externals with subdirectories), but should lessen beginner pain and encourage useage of declare.

<img width="897" alt="screen shot 2017-07-31 at 12 31 10 pm" src="https://user-images.githubusercontent.com/480637/28775290-6e20cbdc-75f1-11e7-8cc5-d73b8396de31.png">

danomatika commented:
Technical background:

* only eited the GUI tcl
* added the global `$::docspath` variable which set the values:
  - empty: not set
  - DISABLED: ignore documents directory feature
  - anything else: checked as a valid path
* the $::docspath variable is stored in the GUI preferences
* added a number of procs for this feature in `pd-gui.tcl`, they are currently global and I'm thinking of putting them in a namespace and possibly in their own file (global is messy enough as it is)
* added a proc to get the initial dialog directory for open/close etc dialogs (home or docspath if it's active & valid)
* adding to the path uses the new "add-to-path" pd message
* deken has been updated to check the $::docspath if it exists, so a deken with these updates should still work on an older Pd without $::docspath

TODOS:

* test that `$::env(HOME)` returns the correct dir (or anything at all) on Windows
* figure out consistency check error after using the new "add-to-path" pd message

megrimm commented:
+1

Wow... just wow

> On Jul 31, 2017, at 7:17 AM, Dan Wilcox <notifications@github.com> wrote:
> 
> Technical background:
> 
> only eited the GUI tcl
> added the global $::docspath variable which set the values:
> empty: not set
> DISABLED: ignore documents directory feature
> anything else: checked as a valid path
> the $::docspath variable is stored in the GUI preferences
> added a number of procs for this feature in pd-gui.tcl, they are currently global and I'm thinking of putting them in a namespace and possibly in their own file (global is messy enough as it is)
> added a proc to get the initial dialog directory for open/close etc dialogs (home or docspath if it's active & valid)
> adding to the path uses the new "add-to-path" pd message
> deken has been updated to check the $::docspath if it exists, so a deken with these updates should still work on an older Pd without $::docspath
> TODOS:
> 
> test that $::env(HOME) returns the correct dir (or anything at all) on Windows
> figure out consistency check error after using the new "add-to-path" pd message
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

fdch commented:
+1 

yes! Very helpful and looks great.

danomatika commented:
I'm currently reworking this as a built-in plugin like deken.

danomatika commented:
Ok, this is a bit cleaner now. I pulled the guts out of pd-gui.tcl and into a pd_docspath.tcl plugin which is loaded after deken and uses it's own namespace.

I also removed some of the extra globals I added previously. The only part I needed to adjust was overriding the `::filenewdir` and `::fileopendir` values if the docspath is enabled. These didn't seem to be used consistently throughout the GUI, so they now are used for most open/save panels.

Also, these default to [pwd] but they are now set to home for all platforms when the GUI is started first and always set to home on Windows as people run the .exe directly. The previous commandline behavior on macOS & Linux remains.

porres commented:
Not sure if this is a bug, 
But what is "clearing" the Externals Install Directory supposed to do? Seems it's doing the same as "reset" (reverting to standard paths as before)

porres commented:
"disabling" the Pd Documents folder gives an error
![unnamed](https://user-images.githubusercontent.com/5360142/28852510-fcda217a-76ff-11e7-96da-5d4a8297cc81.png)

porres commented:
If I choose a different Externals install Directory other than the "externals" subfolder (say "libs" for example) and hit ok, it still creates an "externals" subdirectory and adds it to the search path, whereas "libs" is not added! Though deken still asks me to download to "libs". As I see, it shouldn't create "externals" nor add it to the path, and then add "libs" to the path.

millerpuckette commented:
This looks useful enough that I should wait on releasing 0.46 a few days for this to settle down and then include it.

chr15m commented:
Wow Dan! :+1:

jmej commented:
this is so good!! +100

danomatika commented:
Ok, bugs fixed. New test build Pd-0.48.0-test6-docspath.app.zip: http://docs.danomatika.com/pdbuilds/

porres commented:
> If I choose a different Externals install Directory other than the "externals" subfolder (say "libs" for example) and hit ok, it still creates an "externals" subdirectory and adds it to the search path, whereas "libs" is not added! Though deken still asks me to download to "libs". As I see, it shouldn't create "externals" nor add it to the path, and then add "libs" to the path.

Ok, now, If I choose a different Externals install Directory other than the "externals" subfolder (say "libs" for example) and hit ok, it does not create an "externals" subdirectory anymore. But it  **still** adds it to the search path, even though this folder doesn't even exist whereas "libs" is not added! 

Not that I really wanted to do this, but I'm just trying to twist and turn the interface inside out, to find bugs and weird behaviour.

cheers


<img width="1256" alt="captura de tela 2017-08-04 as 00 10 05" src="https://user-images.githubusercontent.com/5360142/28952706-504ce220-78a9-11e7-9aeb-0d71cf91068d.png">

danomatika commented:
New macOS test build Pd-0.48.0-test6-docspath3.app.zip: http://docs.danomatika.com/pdbuilds/

>But it still adds it to the search path, even though this folder doesn't even exist

I reworked some things and this should be fixed.

>whereas "libs" is not added!

The externals install directory is only added to the search paths automatically when it is the "externals" subdir inside the current documents directory. If you want a custom install dir, you need to add that path manually. This is for those who don't want the root folder added to their paths and don't use folder prefixes when using libs ala [cyclone/match].

danomatika commented:
New macOS test build Pd-0.48.0-test6-docspath4.app.zip: http://docs.danomatika.com/pdbuilds/

millerpuckette commented:
I went ahead and merged into master but made one change - I delayed calling pd_docsdir::init until it's needed (during first deken download), as I found it annoying to get the popup right when starting Pd up fro the first time.

danomatika commented:
I'd argue this is an important feature for beginners as it clues them in to where to start organizing things. The documents directory is both a response to feedback that many students *expect* Pd to use this kind of standard location *and* to the issue of where Deken can install things by default.

Experienced users should only need to hit No on the welcome dialog and it will never show again. If it keeps showing, that is a bug.

> On Aug 8, 2017, at 3:52 AM, Miller Puckette <notifications@github.com> wrote:
> 
> I went ahead and merged into master but made one change - I delayed calling pd_docsdir::init until it's needed (during first deken download), as I found it annoying to get the popup right when starting Pd up fro the first time.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/pure-data/pure-data/pull/152#issuecomment-320827551>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAdVfVVJuld39d4DjJu62uFYif3EAzfqks5sV79FgaJpZM4OoPIX>.
> 

--------
Dan Wilcox
@danomatika <http://twitter.com/danomatika>
danomatika.com <http://danomatika.com/>
robotcowboy.com <http://robotcowboy.com/>

danomatika commented:
> I don't think it's needed if deken isn't used, but I might be wrong about this.

The documents directory is designed to be used without deken. In this case, it creates the default directory and opens file dialogs using that location as a helper for beginners.

> On Aug 8, 2017, at 10:17 AM, Dan Wilcox <danomatika@gmail.com> wrote:
> 
> I'd argue this is an important feature for beginners as it clues them in to where to start organizing things. The documents directory is both a response to feedback that many students *expect* Pd to use this kind of standard location *and* to the issue of where Deken can install things by default.
> 
> Experienced users should only need to hit No on the welcome dialog and it will never show again. If it keeps showing, that is a bug.
> 
>> On Aug 8, 2017, at 3:52 AM, Miller Puckette <notifications@github.com <mailto:notifications@github.com>> wrote:
>> 
>> I went ahead and merged into master but made one change - I delayed calling pd_docsdir::init until it's needed (during first deken download), as I found it annoying to get the popup right when starting Pd up fro the first time.
>> 
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub <https://github.com/pure-data/pure-data/pull/152#issuecomment-320827551>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAdVfVVJuld39d4DjJu62uFYif3EAzfqks5sV79FgaJpZM4OoPIX>.
>> 
> 
> --------
> Dan Wilcox
> @danomatika <http://twitter.com/danomatika>
> danomatika.com <http://danomatika.com/>
> robotcowboy.com <http://robotcowboy.com/>
> 
> 
> 

--------
Dan Wilcox
@danomatika <http://twitter.com/danomatika>
danomatika.com <http://danomatika.com/>
robotcowboy.com <http://robotcowboy.com/>

porres commented:
I find it annoying too, and could get along without it, BUT... I know how important this is for the sake of user friendliness and beginners as well. Moreover, this is only asked once in a lifetime, right? When you upgrade Pd, you don't need to bother again, huh?

cheers

danomatika commented:
That is correct, it's only ever shown if the docspath key in the GUI preferences is empty. It will survive updates and is not connected to the core user settings.

If we're seeing Welcome dialog reopen after either setting the path or disabling it, then there is a bug.

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com


> On Aug 8, 2017, at 7:06 PM, Alexandre Torres Porres <notifications@github.com> wrote:
> 
> I find it annoying too, and could get along without it, BUT... I know how important this is for the sake of user friendliness and beginners as well. Moreover, this is only asked once in a lifetime, right? When you upgrade Pd, you don't need to bother again, huh?
> 
> cheers
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

millerpuckette commented:
I think it's ppropriate for new users who are just concerned about getting stuff running on their own machines to use some sort of automated setup routine.  But there are other situations (such as a university lecture course where students develop patches they can share with others and/or turn in to be marked) where it's important not to unknowingly rely on some abstraction or extern that hapens to be on one's own machine.  For this reason I tell all my students to keep their search paths empty and rely on [declare' objects in the patch for everything.  So I think it's misleading to say that the "proper" way to set up one's Pd installation is to have a directory such as Documents/Pd at all.  I don't mind Pd having a feature that sets such a thing up, but I'd suggest avoiding any suggestion that this is a standard way of doing things -- it's sometimes appropriate and sometimes not.

megrimm commented:
i think there are things i do not know, maybe they are implemented, maybe
discussed (at length of course:), and maybe they are on the table.

what I am asking is what is the "ideal"?

if a student uses an external and issues [declare foo] could deken not just
auto download the lib based on [declare]? granted this is if the lib is in
deken and I assume there there is some way to know what object is in what
lib.

on the list there is another discussion of some kind of "export" function
that finds the externs and "packages" all libs, abs, etc.... maybe this is
"proper" way. hc did this in pdx no?

some other languages are doing this no? i think in node.js one does $npm
install and all modules used get localized (i think... i actually don't
know how this works or really understand) so the "package" is deployable...
(again this is my understanding but may not be true)

I have always just used something like [zexy/any2list] when using a single
object compiled as a binary and only use declare to load full libs like
Gem. i always "thought" this was the proper way but is it? can deken find
instances of [foo/...] where "/" is the delimiter and prior to that is the
lib to download?

but maybe we don't was "auto" download anything! is that bad? we could warn
users that "you are about to download zexy" but then another dialog box!!!

I guess what I am asking, can deken do all this automatically and THAT be
the proper way?

so it seems like to pave a new future might be creating a new "proper way"
but who knows...

to end the ramble here and quote nietzsche: "the future influences the
present as much as the past"

On Wed, Aug 9, 2017 at 1:53 PM, Miller Puckette <notifications@github.com>
wrote:

> I think it's ppropriate for new users who are just concerned about getting
> stuff running on their own machines to use some sort of automated setup
> routine. But there are other situations (such as a university lecture
> course where students develop patches they can share with others and/or
> turn in to be marked) where it's important not to unknowingly rely on some
> abstraction or extern that hapens to be on one's own machine. For this
> reason I tell all my students to keep their search paths empty and rely on
> [declare' objects in the patch for everything. So I think it's misleading
> to say that the "proper" way to set up one's Pd installation is to have a
> directory such as Documents/Pd at all. I don't mind Pd having a feature
> that sets such a thing up, but I'd suggest avoiding any suggestion that
> this is a standard way of doing things -- it's sometimes appropriate and
> sometimes not.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/pure-data/pure-data/pull/152#issuecomment-321332197>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIOr31ghdYECkgT8bxDiA01QAczRr9_ks5sWfIqgaJpZM4OoPIX>
> .
>



-- 
____________________
m.e.grimm, m.f.a, ed.m.
syracuse u., tc3
megrimm.net
____________________

porres commented:
Hi, about using [declare], I wrote about it on my tutorial on how to install and manage externals. But I have to say this only works well if you use "standard paths" and the -stdlib and -stdpath flags. Hence all that discussion we had in my other pull requests. 

Say you have externals elsewhere than Standard Paths (which we now made a sharp decision to move towards to), then if you wanna use declare to manage this, you need to use relative or absolute paths, which I think are much more complicated to manage. 

This goes back to me pointing we need to do yet more things to make the "path mechanism" behave like people expect it to. So I take this opportunity to say that we could make the "-path" flag in [declare] also search and find subdirectories in user added paths. Let me illustrate.

If I have cyclone in a standard path, I can keep the search path empty and just use declare to call it. This seems to be what Miller suggests it's a good practice, and one that I also write about in my tutorial.

See the picture below: search path is empty, and declare is being used to load cyclone with the "-stdpath" flag.

<img width="717" alt="captura de tela 2017-08-09 as 15 28 48" src="https://user-images.githubusercontent.com/5360142/29137682-85ada536-7d17-11e7-9bda-447dca5e092a.png">

So, now we do not want to mess with standard paths no more, and we can put it in a different folder, such as ~/Documents/Pd/externals - if I have cyclone in there, I cannot use [declare] in a simple way like this: 

<img width="880" alt="captura de tela 2017-08-09 as 15 33 35" src="https://user-images.githubusercontent.com/5360142/29137866-2446d852-7d18-11e7-97a2-21c394a31b66.png">

Unless I do an absolute path, and that is much more complicated than before with standard paths. Maybe now I can make it finally clear a major difference that I've been long trying to point out.

One solution we could have is making [declare -path] also search in the added search paths.

It's funny that all this discussion still makes me think a simple solution was to just change the standard path to a new location :) - I already gave up on it, and joined the team to work on a new solution. But all this discussion makes me wonder we still need to have an understanding and a consensus on this mechanism. 

Anyway, hope I'm making sense now

cheers

porres commented:
Now, regarding the discussion of initially asking to add Documents/Pd/Externals to the seach path with a one-time only dialog, not that this is not where you add each subdirectory to the search path, this happens elsewhere, when deken finishes downloading an external package, where it then asks you if you also want to add it to the search path. Maybe you don't wanna add it so you can manage things with [declare], so you just say "no" at that moment.

But my point is that it is still important and necessary to add this new folder we now want to be the "new way" instead of "standard paths".

danomatika commented:
> For this reason I tell all my students to keep their search paths empty and rely on [declare' objects in the patch for everything. 

They can easily disable the documents directory then...

danomatika commented:
Or the whole plugin can be an external plugin.

porres commented:
> I have always just used something like [zexy/any2list] when using a single object compiled as a binary and only use declare to load full libs like Gem. i always "thought" this was the proper way but is it?

I now just believe there's not a one single "proper way" of doing this, at least not a consensus. There are different options and the software wasn't designed to impose one "proper" alternative. Like, I always though using "standard paths" was the "proper" way to do things, but what a surprise...

Anyway, yes, "slash" declaration is also one way of doing things. What I did when I wrote my tutorial is that I mapped "every way" possible, making it clear the differences, and now, with all the discussion and these new things coming up, I have to say I still do not know how to revise this work. I still don't know how to mention and discuss the usage of standard paths, for instance... cause it seems it's just something to throw away now, since it's very complicated and we don't wanna fix it, just do something else.

So, I keep talking about this tutorial, and here it is anyway, if you missed it: https://drive.google.com/file/d/0B3AoiT0xk8fnYWV5anJZeXZ2MDg/view

danomatika commented:
Also, before this devolves further and further into semantics, I feel the work that I have done solves three problems:

* Cluing-in beginners where to work with Pd through emulating other environments such as Processing by providing a mechanism to create and utilize a Documents/Pd directory.
* Also providing a place to download externals via deken in a centralized location, again within the Documents/Pd directory.
* Provides the ability to add the external to the user search paths.

Let's say there are two camps:

* Group1 -> new tools: For educators and students working with Processing & Arduino, the Documents/Pd location is an intuitive "proper place." No as comfortable with creating their own directories and organization.
* Group 2-> old tools: Comfortable with creating their own directories and organization.

I feel I have provided the tools for Group 1 to create this place while at the same time allow Group 2 to totally disable it. You can still do what you want and providing a small helper dialog to tell people about this choice is important.

If Group 2 wants to work completely within their own organization, so being it. Say "No" to the dialog and or disable in the Paths preferences. Note that if you enable the Documents/Pd directory you can still choose a manual location to download the externals to using deken *each* time. Again, I feel I have balanced the needs of both Groups by adding new functionality but trying not to impede the desired workflow of each.

Please reinstate the welcome dialog in #168. It make no mention of "proper this or that."

danomatika commented:
> it's important not to unknowingly rely on some abstraction or extern that hapens to be on one's own machine

I say that's a fault with the author of the patch who should include a list of required externals in a comment or readme. Deken and my Documents Directory work are meant to help move people away from the Extended kitchen sink model. We're missing proper package management with dependencies and all of that, but we can at least smooth over the instructions on how to install externals as, right now, it involves deeper understanding in the path mechanism, search paths, and [declare].

porres commented:
yes, I think it's a good solution "in between", with some minimal compromise for these two group of people, and I'm taking it.

But I still think much debate comes from different views and not fully understanding what is the mechanism to install and manage externals we have today, and where we can go from it - so I think it'd be good to sort out "proper ways of doing stuff" later on, and perhaps make further enhancements to benefit both new/old tools groups. Sorry I'm already pointing up further suggestions, I know it's not the time. 

Regarding this matter, all I can say is that I hope the welcome dialog is brought back, otherwise it ends up being kinda worthless and not a good solution in between again. And having it as an external plug in would be the worst case scenario... I also think there is a consensus in the Pd Community that this new feature is really helpful as it is.

millerpuckette commented:
I'll be teaching a workshop early Sept and perhaps that will give me a chance to understand better what is being argued here.. until then I'm unwilling to commit to throwing up a welcome window - it's not exactly a logical choice, it's simply that I personally find that behavior very irritating. Perhaps this is the view of a minority of one.  It looks to me as if there are two things being done, a "welcome to Pd" gesture and an offer to start helping the user manage patches.  Perhaps a welcome-to-Pd gesture is warranted, but then I think it should point the user at the most frequently used resources rather than a yes-no-cancel choice.  (incidentally, "cancel" seems to mean, "throw this welcome up again next time").

OTOH, perhaps as Alexandre says, the dialog is now too well hidden.  I was under the impression that the new features in the path dialog are an alternative way to accomplish the same thing (establish a user-specific path and optionally create a directory for it).  But perhaps there needs to be a more prominent place the user can go to call pd_docsdir::init without bugging the user ala microsoft clippy :)

danomatika commented:
>it's simply that I personally find that behavior very irritating

It's only ever shown once. If you make a choice, you don't see it again. I'd understand it being irritating if it opened every time you open Pd, but this is not the case.

If it is opening overtime Pd is open, that is a bug.

>(incidentally, "cancel" seems to mean, "throw this welcome up again next time").

It is, as in "I haven't decided yet." The cancel option can be easily removed.

>the dialog is now too well hidden

Yes.

danomatika commented:
Also pd_docsdir::init does more than throw up the dialog, it also makes sure that, if the path is set, it exists and it sets the panel initialdir values used within the GUI. It needs to be called when Pd is started for this to work.

If this behavior is not wanted, I can make changes but I would have hoped these issues would come up *before* a merge.

millerpuckette commented:
yes - I put the changes in before merging but then didn't know how to publish a branch without risking pulverizing something in the repo.  I assumed my changes were minor enough not to cause any trouble although I guess not.

Is there a next-best-thing besides throwing the dialog up on Pd's startup?  From the comments above, it looks like it would be useful to have an easy-to-find way for the user to call ::pd_docsdir::init, outside of teh context of first-setting-Pd-up -- even if it does get displayed on first-running-Pd, I would want to be able to get back to it later.  For instance, should this be in the preferences submenu?

danomatika commented:
>yes - I put the changes in before merging but then didn't know how to publish a branch without risking pulverizing something in the repo.

One method is to add the PR source as a git remote which you can then pull from to test things without merging. The other method is to do a separate clone.

>Is there a next-best-thing besides throwing the dialog up on Pd's startup?

My thinking was that we did not want to create the folders automatically, so we simply ask first. At the same time, this clues in beginners about the folder. If it's hidden in a menu, they will most likely not enable it or even know about it.

>From the comments above, it looks like it would be useful to have an easy-to-find way for the user to call ::pd_docsdir::init, outside of teh context of first-setting-Pd-up

It needs to be called at start for the path checking behavior and open panel initialdir stuff, as previously mentioned. Those checks could be split into a separate proc, but I don't think that's needed as what happens in there is very simple:

1. check if the documents directory is set (aka loaded from the GUI prefs)
  * if it's not set, try to detect a suitable location and throw up the dialog and ask the user if they want 
to create the docs directory there. If the final location is not writable, disable the docspath feature and return.
  * if it is already set, check if the location exists, if not then ask them to select a new location
  * if it is disabled aka == "DISABLED", do nothing further and return
2. set the $::fileopen & $::filenew paths to the current documents path (if valid), this way beginners are encouraged to save things in the new directory, matching behavior for Processing etc. This in no way keeps them from saving and opening for any other location, just sets the start location.

>even if it does get displayed on first-running-Pd, I would want to be able to get back to it later. For instance, should this be in the preferences submenu?

The configuration for the directory is in the Path dialog. You can change the location or disable it there. Same with the deken install path.

porres commented:
> My thinking was that we did not want to create the folders automatically, so we simply ask first. 

Yeah, my first choice was that it'd automatically be added, but there was this resistance to do so. Asking about it seems like a reasonable compromise and a solution to the previous resistance. I wouldn't know of another solution, as it would narrow down to keep manually doing things and knowing about them beforehand.

millerpuckette commented:
So it's at least clear that, regardless of the  desirability of throwing up the "welcome" dialog, my commit
to delay doing so until deken gets called was wrong in other ways -- I'll undo that, but still would like in the future to find a solution we're all comfortable with - I remain not-comfortable with the status quo.  In the meantime though, a simple suggestion - replace the "cancel" label with "ask me later".

danomatika commented:
I can also remove the "Welcome to Pd" text. I mainly added that since it seems kind of dry with only the question.

millerpuckette commented:
... i do think it's better without the "welcome..." but this might be hypersensitivity on my part - it feels patronizing, and anyway, unlike the roadside sign "welcome to Kansas", somehow, I didn't just walk or drive inside Pd :)

danomatika commented:
I've removed the Welcome line in #168. I' suggested further comments move there.

I can look into changing the button text, but that requires creating a custom dialog since we're currently using "yesnocancel" option of tk_messageBox whose defaults cannot be changed.

millerpuckette commented:
Maybe it's not worth it then to fix "cancel" string.

danomatika commented:
Yeah, at the time I had the same thought of better button labels, but didn't feel like making a custom, possibly non-standard dialog. Trade-off.
